flow/display_list_benchmarks.cc

```diff
@@ -0,0 +1,120 @@
+#include "flutter/fml/mapping.h"
+#include "flutter/testing/test_metal_context.h"
+#include "flutter/testing/test_metal_surface.h"
+#include "flutter/testing/testing.h"
+
+#include "third_party/skia/include/core/SkPoint.h"
+#include "third_party/skia/include/core/SkSurface.h"
+
+namespace flutter {
+
+namespace testing {
+
+class CanvasProvider {
+ public:
+  virtual const std::string BackendName() = 0;
+  virtual sk_sp<SkSurface> GetSurface() = 0;
+
+  virtual bool Snapshot(std::string filename) {
+    auto image = GetSurface()->makeImageSnapshot();
+    if (!image) {
+      return false;
+    }
+    auto raster = image->makeRasterImage();
+    if (!raster) {
+      return false;
+    }
+    auto data = raster->encodeToData();
+    if (!data) {
+      return false;
+    }
+    fml::NonOwnedMapping mapping(static_cast<const uint8_t*>(data->data()),
+                                 data->size());
+    return WriteAtomically(OpenFixturesDirectory(), filename.c_str(), mapping);
+  }
+};
+
+class SoftwareCanvasProvider : public CanvasProvider {
+ public:
+  SoftwareCanvasProvider(const size_t width, const size_t height) {
+    surface_ = SkSurface::MakeRasterN32Premul(width, height);
+    surface_->getCanvas()->clear(SK_ColorTRANSPARENT);
+  }
+
+  sk_sp<SkSurface> GetSurface() override { return surface_; }
+
+  const std::string BackendName() override { return "Software"; }
+
+ private:
+  sk_sp<SkSurface> surface_;
+};
+
+class MetalCanvasProvider : public CanvasProvider {
+ public:
+  MetalCanvasProvider(const size_t width, const size_t height) {
+    metal_context_ = std::make_unique<TestMetalContext>();
+    metal_surface_ =
+        TestMetalSurface::Create(*metal_context_, SkISize::Make(width, height));
+    metal_surface_->GetSurface()->getCanvas()->clear(SK_ColorTRANSPARENT);
+  }
+
+  sk_sp<SkSurface> GetSurface() override {
+    return metal_surface_->GetSurface();
+  }
+
+  const std::string BackendName() override { return "Metal"; }
+
+ private:
+  std::unique_ptr<TestMetalContext> metal_context_;
+  std::unique_ptr<TestMetalSurface> metal_surface_;
+};
+
+template <class T>
+void DrawLine(benchmark::State& state) {
+  DisplayListBuilder builder;
+  T canvas_provider(state.range(0), state.range(0));
+
+  auto canvas = canvas_provider.GetSurface()->getCanvas();
+
+  // Always draw 10k lines regardless of the line length requested
+  size_t length = state.range(0);
+  for (size_t i = 0; i < 10000; i++) {
+    // The (approximate) line lengths are passed to us via the benchmark state
+    // object, so that we can benchmark the performance of drawLine and how it
+    // differs based on line length
+    builder.drawLine(SkPoint::Make(i % length, 0),
+                     SkPoint::Make(length - i % length, length));
+  }
+
+  auto displaylist = builder.Build();
+
+  // We only want to time the actual rasterization.
+  while (state.KeepRunning()) {
+    displaylist->RenderTo(canvas);
+    canvas->flush();
```

Comment:
In here you'll wnat to fish out the GrDirectContext and call `flushAndSubmit` or `flush`/`submit` separately.